### PR TITLE
Filter posts by tag

### PR DESCRIPTION
**Pull request by chloealee**
#### What's this PR do?
This PR allows adds the `filter[tag]=[tag_slug]` to query the `/posts` endpoint by a certain tag. 
With this, we can now filter posts by a specific tag. 

#### How should this be reviewed?
On the campaign single view page, try to filter by a tag! Make sure the responses are what you expect (posts with that specific tag). 

#### Relevant tickets
Fixes https://www.pivotaltracker.com/n/projects/2019429

#### Checklist
- [ ] Documentation added for new features/changed endpoints.
- [ ] Tested on staging.
- [ ] Pinged a PM if this is a larger PR that would benefit from some additional testing love.

### CODE REVIEW

File: app/Models/Post.php [hunk 1]
```diff
@@ -203,12 +203,22 @@ public function untag($tagName)
...
     {
         return $query->whereDoesntHave('tags', function ($query) use ($tagSlug) {
             $query->where('tag_slug', '=', $tagSlug);
         });
     }
+
+    /**
+     * Returns posts with specific tag.
+     */
+    public function scopeWithTag($query, $tagSlug)
```
Comment (sbsmith86) [inline app/Models/Post.php]: Nice!

Comment (sbsmith86) [reply app/Models/Post.php]: Should we have it be `withTags()` and instead of checking for equality, we check if the `tag_slug` is in a list of tags?

File: app/Models/Post.php [hunk 2]
```diff
@@ -203,12 +203,22 @@ public function untag($tagName)
     }
 
     /**
-     * Remove the given tag from this post.
+     * Returns posts without specific tag.
```
Comment (katiecrane) [inline app/Models/Post.php]: Great catch!!!

File: resources/assets/components/CampaignSingle/index.js [hunk 3]
```diff
@@ -24,15 +24,37 @@ class CampaignSingle extends React.Component {
...
     this.getPostsByStatus('accepted', this.props.campaign.id);
   }
 
-  // Filter posts based on status.
-  filterPosts(status) {
-    this.getPostsByStatus(status.toLowerCase(), this.props.campaign.id);
+  // Filter posts based on status or tag(s).
+  filterPosts(filter) {
+    // If the filter is a status, make API call to get posts by status.
+    if (filter === 'Pending' || filter === 'Accepted' || filter === 'Rejected') {
```
Comment (sbsmith86) [inline resources/assets/components/CampaignSingle/index.js]: You can use [`include()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes) here to see if the filter is in an array of values, instead of multiple or statements

File: resources/assets/components/CampaignSingle/index.js [hunk 4]
```diff
@@ -24,15 +24,37 @@ class CampaignSingle extends React.Component {
...
+      this.getPostsByTag(tagSlug, this.props.campaign.id);
+    }
+
+  }
+
+  // Create tag slug
+  createTagSlug(filter) {
+    var splitFilter = filter.toLowerCase().split(' ');
+
+    if (splitFilter[splitFilter.length-1] === '👻') {
```
Comment (katiecrane) [inline resources/assets/components/CampaignSingle/index.js]: Hahaha I'd be on board with removing the 👻  from the dropdown in order to not have to do this. But at the same time, I love this line.

File: app/Models/Post.php (hunk 1 above)
Comment (chloealee) [reply app/Models/Post.php]: I was thinking that too and like that idea! For filtering by tag, you can only filter a tag one at a time (from the drop down menu) so we don't need it for this but could be good for the api in general to be able to send in multiple tags!

**Comment by chloealee** [comment]
@katiecrane I did run `composer update` after I rebased for the tags work!

**Comment by chloealee** [comment]
ah but maybe shouldn't have done that since it should have already updated with your merge? @katiecrane

File: resources/assets/components/CampaignSingle/index.js [hunk 5]
```diff
@@ -24,15 +24,37 @@ class CampaignSingle extends React.Component {
...
 
-  // Filter posts based on status.
-  filterPosts(status) {
-    this.getPostsByStatus(status.toLowerCase(), this.props.campaign.id);
+  // Filter posts based on status or tag(s).
+  filterPosts(filter) {
+    // If the filter is a status, make API call to get posts by status.
+    if (filter === 'Pending' || filter === 'Accepted' || filter === 'Rejected') {
+      this.getPostsByStatus(filter.toLowerCase(), this.props.campaign.id);
+    } else {
```
Comment (sbsmith86) [inline resources/assets/components/CampaignSingle/index.js]: What happens if we want to filter by status AND tag. Could we potentially grab all the filter fields and then make one request with all of the filter values? If this feels out of scope for this PR, we could create a card to update the functionality. But we should be clear that you can only do one or the other.

**Comment by katiecrane** [comment]
@chloealee that sounds fine to me!

File: resources/assets/components/PostFilter/index.js [hunk 6]
```diff
@@ -21,6 +21,11 @@ class PostFilter extends React.Component {
                   <option>Accepted</option>
                   <option>Pending</option>
                   <option>Rejected</option>
+                  <option>Good Photo</option>
```
Comment (sbsmith86) [inline resources/assets/components/PostFilter/index.js]: I think we should do something more like `<option value="good-photo">Good Photo</option>` This way you don't have to do the `createTagSlug()` work, because you are already passing the slug as the value.

File: resources/assets/components/CampaignSingle/index.js (hunk 5 above)
Comment (sbsmith86) [reply resources/assets/components/CampaignSingle/index.js]: Ah I see that below, we just do one large select list with filter options so we only allow one filter. In this case it is fine, but I definitely think we should update this functionality so that the `status` filter and `tags` filter are separate. We should be able to do something like give me all `Accepted` posts that have a `good-photo` tag. So can you make a note to make a new card for this work and we can prioritize that later?

File: resources/assets/components/CampaignSingle/index.js (hunk 3 above)
Comment (chloealee) [reply resources/assets/components/CampaignSingle/index.js]: oh awesome! So much cleaner.

File: resources/assets/components/CampaignSingle/index.js (hunk 5 above)
Comment (chloealee) [reply resources/assets/components/CampaignSingle/index.js]: Yes - I'll make a card for this now! cc @ngjo

File: resources/assets/components/PostFilter/index.js (hunk 6 above)
Comment (chloealee) [reply resources/assets/components/PostFilter/index.js]: ohhhhh yes!

File: app/Models/Post.php (hunk 1 above)
Comment (sbsmith86) [reply app/Models/Post.php]: Yea and it looks like we only allow you to filter out one tag. so I think both of these should be updated. Will make a card!

Comment (chloealee) [reply app/Models/Post.php]: awesome, thanks Shae!

File: resources/assets/components/PostFilter/index.js [hunk 7]
```diff
@@ -18,9 +18,14 @@ class PostFilter extends React.Component {
           <h4>Post Filter</h4>
           <div className="select">
               <select onChange={(event) => this.change(event)}>
```
Comment (sbsmith86) [inline resources/assets/components/PostFilter/index.js]: Looks great! Let's make note for future filtering work that we should create a `<Filter>` component that takes in a array of value/labels` and renders the select list and handles the change event. So as we build out multiple filters, we don't have to hard code values and they can be more flexible.

Comment (chloealee) [reply resources/assets/components/PostFilter/index.js]: love it! I'll add a `@TODO` comment now.
